lib/Dialect/TritonGPU/Transforms/OptimizeThreadLocality.cpp

```diff
@@ -1,20 +1,95 @@
 #include "mlir/Pass/Pass.h"
 #include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
 #include "mlir/Transforms/Passes.h"
 #include "triton/Dialect/TritonGPU/IR/Dialect.h"
 #include "triton/Dialect/TritonGPU/Transforms/Passes.h"
 #include "triton/Dialect/TritonGPU/Transforms/Utility.h"
 #include <memory>
+#include <numeric>
 #define GEN_PASS_CLASSES
 #include "triton/Dialect/TritonGPU/Transforms/Passes.h.inc"
 
 using namespace mlir;
+namespace {
+// Change the layout of view destination when used by a reduction in order to
+// minimize the amount of cross thread communication for the reduction.
+struct OptimizeViewLayoutPattern
+    : public mlir::OpRewritePattern<triton::ReshapeOp> {
+  OptimizeViewLayoutPattern(mlir::MLIRContext *context)
+      : OpRewritePattern<triton::ReshapeOp>(context, 1) {}
+
+  mlir::LogicalResult
+  matchAndRewrite(triton::ReshapeOp viewOp,
+                  mlir::PatternRewriter &rewriter) const override {
+    if (!viewOp.getAllowReorder())
+      return failure();
+    std::optional<int> reductionAxis;
+    for (Operation *user : viewOp.getResult().getUsers()) {
+      if (auto reduceOp = dyn_cast<triton::ReduceOp>(user)) {
+        if (reductionAxis) {
+          if (reductionAxis != reduceOp.getAxis())
+            return failure();
+        } else {
+          reductionAxis = reduceOp.getAxis();
+        }
+      }
+    }
+    if (!reductionAxis)
+      return failure();
+    auto tensorType = viewOp.getResult().getType().cast<RankedTensorType>();
+    ArrayRef<int64_t> shape = tensorType.getShape();
+    llvm::SmallVector<unsigned> order;
+    for (int i = shape.size() - 1; i >= 0; i--) {
+      if (i != *reductionAxis)
+        order.push_back(i);
+    }
+    // Make the reduction axis last so that elements won't be distributed
+    // amongst threads along this dimension.
+    order.push_back(*reductionAxis);
```

Comment:
We don't care about the order of the other dims, we just want the reduction axis to be last.  But this forces all the other dims to be in a specific order.

If the input layout is (say) `2, 0, 1` and we want to reduce along dimension 1, then shouldn't we keep that layout, rather than forcing to `2, 1, 0`?

Reply:
good point, I changed the logic to start with the original order.